Add `FixedPointOperand` blanket implementation

juangirini opened the pull request:
This PR replaces the explicit `FixedPointOperand` impls by a blanket one so that `Balance` implements it.

juangirini committed:
bound `Balance` to `FixedPointOperand`

juangirini commented:
bot fmt

Commit:
".git/.scripts/commands/fmt/fmt.sh"

juangirini committed:
clean up code

gavofyork commented on frame/assets/src/types.rs:
```diff
@@ -293,8 +293,8 @@ where
 	CON: Convert<BalanceOf<F, T>, AssetBalanceOf<T, I>>,
-	BalanceOf<F, T>: FixedPointOperand + Zero,
-	AssetBalanceOf<T, I>: FixedPointOperand + Zero,
+	BalanceOf<F, T>: Zero,
+	AssetBalanceOf<T, I>: Zero,
```
```suggestion
```

gavofyork commented on frame/assets/src/lib.rs:
```diff
@@ -241,7 +241,8 @@ pub mod pallet {
 			+ MaybeSerializeDeserialize
 			+ MaxEncodedLen
-			+ TypeInfo;
+			+ TypeInfo
+			+ FixedPointOperand;
```
```suggestion
			+ TypeInfo;
```

gavofyork replied:
This should not be needed. `FixedPointOperand` has a blanket impl and all bounds required for it are already bounded by `Copy + AtLeast32BitUnsigned`.

gavofyork reviewed (changes requested):
Explicitly stating the bound should not be needed.

juangirini committed:
Apply suggestions from code review

Co-authored-by: Gavin Wood <gavin@parity.io>

juangirini committed:
wip

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3264496

juangirini committed:
add blanket `FixedPointOperand` impl

juangirini commented:
I haven't found that blanket impl so I added it as part of this PR. Also, `Copy + AtLeast32BitUnsigned` are missing `CheckedNeg` compared to `FixedPointOperand` so I added that too.

athei commented on frame/nis/src/lib.rs:
```diff
@@ -224,7 +226,8 @@ pub mod pallet {
 			+ TypeInfo
-			+ MaxEncodedLen;
+			+ MaxEncodedLen
+			+ CheckedNeg;
```
Why is this necessary? This PR does not remove any bounds in this pallet. Additionally, this type seems to have a lot of repeated bounds over `trait Balance` through which it is already bound in the `type Currency` above.

juangirini replied:
`CurrencyBalance` is just the `Balance` type + `From<u64>` and therefore it needs to implement `Balance`. Todo it through its blanket impl now we need, the newly added, `CheckedNeg`.
Alternatively, to make it more clear, I guess we could just do 
```rust
type CurrencyBalance: Balance + From<u64>;
```

athei replied:
Got it. Yes. I think your suggestion is good. It would bring it more in line with the doc on it.

juangirini committed:
update nis CurrencyBalance

gavofyork commented on frame/assets/src/lib.rs:
```diff
@@ -241,7 +243,8 @@ pub mod pallet {
 			+ MaxEncodedLen
-			+ TypeInfo;
+			+ TypeInfo
+			+ CheckedNeg;
```
Avoid adding here - just add it into `BaseArithmetic` trait/blanket impl along with all the other `Checked*` bounds.

gavofyork reviewed (changes requested):
Getting there. Left a comment.

juangirini committed:
remove CheckedNeg bound

gavofyork reviewed (approved):
LGTM.

juangirini commented:
bot merge
